src/rt/test/func/external_event_source/external_event_source.cc

```diff
@@ -0,0 +1,117 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+#include <test/harness.h>
+#include <verona.h>
+
+using namespace verona::rt;
+
+int buffer[100];
+
+struct Poller : VCown<Poller>
+{
+  int fd;
+  std::atomic<bool> should_notify;
+  int empty_count;
+  Noticeboard<int> buffer_idx;
+  int read;
+
+  Poller() : should_notify(false), empty_count(0), buffer_idx(0), read(0) {}
+
+  void main()
+  {
+    int val, read_old;
+    auto& alloc = ThreadAlloc::get();
+
+    read_old = read;
+    while (read <= buffer_idx.peek(alloc))
+    {
+      val = buffer[read++];
+
+      Systematic::cout() << val << std::endl;
+
+      if (val == 19)
+        return;
+    }
+    if (read > read_old)
+      empty_count = 0;
+    else
+      empty_count++;
+
+    if (empty_count < 10)
+      schedule_lambda(this, [=]() { main(); });
+    else
+    {
+      // add external source and enable notifications
+      Scheduler::add_external_event_source();
+      should_notify = true;
```

Comment:
After `should_notify` you need to recheck the `Noticeboard` to see if new data has arrived, otherwise you can go to sleep here, and never be woken up.